[JIT] check for object equality in constant pooling

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31800 [JIT] check for object equality in constant pooling**
* #31501 [JIT] Add AliasDb API For Changing Aliasing

If we know that two constants are the same object, we can ignore other constraints and pool them together. This fixes an issue introduced by the other PR where quantization relied on constant pooling happening for correctness.

Differential Revision: [D19269499](https://our.internmc.facebook.com/intern/diff/D19269499)

Commit:
check for object equality in constant pooling

[ghstack-poisoned]

jerryzh168 reviewed (approved):
Looks good to me

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 7103c3ff:


* **2/2** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=7103c3ff24e8392b1d1e74213fa0ab0ff5759ed3).

### :detective: 2 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4110262) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_simple_test (1/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79572318) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79572318))
<details>
<summary>
<code>Jan 02 23:55:28 RuntimeError: test_jit_simple failed!</code>
</summary>

```
Jan 02 23:55:28 Ran 2313 tests in 73.307s 
Jan 02 23:55:28  
Jan 02 23:55:28 FAILED (errors=1, skipped=82, expected failures=1) 
Jan 02 23:55:28  
Jan 02 23:55:28 Generating XML reports... 
Jan 02 23:55:28 Traceback (most recent call last): 
Jan 02 23:55:28   File "test/run_test.py", line 456, in <module> 
Jan 02 23:55:28     main() 
Jan 02 23:55:28   File "test/run_test.py", line 449, in main 
Jan 02 23:55:28     raise RuntimeError(message) 
Jan 02 23:55:28 RuntimeError: test_jit_simple failed! 
Jan 02 23:55:28 + cleanup 
Jan 02 23:55:28 + retcode=1 
Jan 02 23:55:28 + set +x 
Jan 02 23:55:28 =================== sccache compilation log =================== 
Jan 02 23:55:28 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 02 23:55:28 Compile requests                 2 
Jan 02 23:55:28 Compile requests executed        0 
Jan 02 23:55:28 Cache hits                       0 
Jan 02 23:55:28 Cache misses                     0 
Jan 02 23:55:28 Cache timeouts                   0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4110263) pytorch_linux_xenial_py3_6_gcc5_4_ge_config_legacy_test (2/2)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=79572696) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=79572696))
<details>
<summary>
<code>Jan 02 23:56:12 RuntimeError: test_jit_legacy failed!</code>
</summary>

```
Jan 02 23:56:12 Ran 2313 tests in 121.889s 
Jan 02 23:56:12  
Jan 02 23:56:12 FAILED (errors=1, skipped=65, expected failures=1) 
Jan 02 23:56:12  
Jan 02 23:56:12 Generating XML reports... 
Jan 02 23:56:12 Traceback (most recent call last): 
Jan 02 23:56:12   File "test/run_test.py", line 456, in <module> 
Jan 02 23:56:12     main() 
Jan 02 23:56:12   File "test/run_test.py", line 449, in main 
Jan 02 23:56:12     raise RuntimeError(message) 
Jan 02 23:56:12 RuntimeError: test_jit_legacy failed! 
Jan 02 23:56:12 =================== sccache compilation log =================== 
Jan 02 23:56:12 + cleanup 
Jan 02 23:56:12 + retcode=1 
Jan 02 23:56:12 + set +x 
Jan 02 23:56:12 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 02 23:56:12 Compile requests               250 
Jan 02 23:56:12 Compile requests executed        0 
Jan 02 23:56:12 Cache hits                       0 
Jan 02 23:56:12 Cache misses                     0 
Jan 02 23:56:12 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "[JIT] check for object equality in constant pooling"


If we know that two constants are the same object, we can ignore other constraints and pool them together. This fixes an issue introduced by the other PR where quantization relied on constant pooling happening for correctness.

Differential Revision: [D19269499](https://our.internmc.facebook.com/intern/diff/D19269499)

[ghstack-poisoned]

jerryzh168 commented on torch/csrc/jit/constants.cpp:
```diff
@@ -176,7 +176,7 @@ RegisterOperators reg({
 
 c10::optional<IValue> toIValue(const Value* v) {
-  if (v->node()->kind() != prim::Constant) {
+  if (v->node()->kind() != prim::Constant || v->type()->cast<FunctionType>()) {
```
is there constant Function nodes?

eellison replied:
Yea

Commit:
Update on "[JIT] check for object equality in constant pooling"


If we know that two constants are the same object, we can ignore other constraints and pool them together. This fixes an issue introduced by the other PR where quantization relied on constant pooling happening for correctness.

Differential Revision: [D19269499](https://our.internmc.facebook.com/intern/diff/D19269499)

[ghstack-poisoned]

Commit:
Update on "[JIT] check for object equality in constant pooling"


If we know that two constants are the same object, we can ignore other constraints and pool them together. This fixes an issue introduced by the other PR where quantization relied on constant pooling happening for correctness.

Differential Revision: [D19269499](https://our.internmc.facebook.com/intern/diff/D19269499)

[ghstack-poisoned]
